### PR TITLE
Add Streaming Dataset

### PR DESCRIPTION
**Pull request by fracapuano**
Part of PR stack

👀 https://github.com/huggingface/lerobot/pull/1173
👉 https://github.com/huggingface/lerobot/pull/1165
👀 https://github.com/huggingface/lerobot/pull/969

## What this does

This PR introduces two key components:
1. A new `StreamingLeRobotDataset` class for efficient streaming of robot datasets
2. A profiling script to measure and analyze its performance characteristics

## `StreamingLeRobotDataset`

The `StreamingLeRobotDataset` class extends the standard dataset functionality with streaming capabilities, allowing data to be processed without loading the entire dataset into memory.

### Key Features:
- **Memory-efficient streaming**: Loads data on-demand rather than all at once
- **Buffer-based sampling**: Maintains a bounded buffer of recent items for efficient access
- **Sharded loading**: Supports loading data from multiple shards for improved performance and randomness
- **Video frame decoding**: Efficiently handles video data downloading only certain frames in a video. Uses caching for improved performance (~10x)
- **Random sampling**: Implements efficient random sampling from the stream

## Profiling Script

The accompanying profiling script (`profile_streaming_dataset.py`) provides detailed performance analysis. **Note: `profile_streaming_dataset.py` is not intended for merge and will be removed when this PR is ready to merge)**

### Features:
- **Detailed timing measurements**: Per-sample and per-run timing statistics
- **Warmup support**: Configurable warmup iterations to measure overhead
- **Line profiling**: Detailed line-by-line profiling of key dataset methods
- **Progress visualization**: Uses tqdm to show progress during profiling runs
- **Comprehensive reporting**: Saves results to a file for further analysis

### Streaming Dataset:
```python
from lerobot.common.datasets.streaming_dataset import StreamingLeRobotDataset

dataset = StreamingLeRobotDataset(
    repo_id="lerobot/aloha_mobile_cabinet",
    buffer_size=1000,
    max_num_shards=16
)

for item in dataset:
    # Process item
    pass
```

### Profiling:
```bash
python -m lerobot.common.datasets.profile_streaming_dataset \
    --repo-id lerobot/aloha_mobile_cabinet
```

## How to checkout & try? (for the reviewer)
Checkout to this branch and use the aforementioned snippets. Unsurprisingly, one needs a tiny bit of version management to make sure everything works 😅 In particular, the latest version of `torchcodec` (enabling single-frame streaming) requires specific dependancies, namely you might need to reinstall/update `torch`, `torchvision`, `torchaudio` and `ffmpeg`. Here are the versions I am using to run this locally:

```bash
torch                                2.7.0
torchaudio                           2.7.0
torchcodec                           0.4.0
torchvision                          0.22.0
imageio                              2.37.0  # ffmpeg
```

cc. @lhoestq for review and @Cadene for viz

### CODE REVIEW

**Comment by fracapuano** [comment]
# Assessing performance

Obtaining an `IterableDataset` that is as close as possible to its non-streaming `Dataset` counterpart requires implementing two main features:

1. **Fast** access to rows when iterating over it (~high throughput when batching the iterableDataset)
2. **Randomized** access to rows (to guarantee eventual ML training happens under the usual i.i.d. assumption)

Both these features are implemented via the `__iter__` method of `StreamingLeRobotDataset`. In particular,
1. **Fast access to the rows content** is obtained using the latest `torchcodec` release, allowing to stream frames from a remote video without downloading the entire video first. Check out [the official documentation](https://docs.pytorch.org/torchcodec/stable/generated_examples/file_like.html#:~:text=An%20obvious%20question%20is:%20which%20is%20faster?%20The%20code%20below%20tests%20that%20question.) for an in-detail explanation of the benefits
2. **Randomization** is achieved using a local buffer that is randomly accessed. The buffer is populated selecting a dataset's shard at random. This 2-level randomization ensures random-enough access to the elements of the dataset (see below for an analysis of this)

## Profiling results ⏱️ 
```bash
TIMING STATISTICS
============================================================
Warmup timing (ms):
  Mean: 7802.74 ± 0.00
  Median: 7802.74
  Range: [7802.74, 7802.74]

Per-sample timing (ms):
  Mean: 10.26 ± 154.29
  Median: 6.00
  Range: [1.89, 7426.07]

Per-run timing (seconds):
  Mean: 28.59 ± 0.70
  Range: [27.38, 29.67]

Throughput:
  Samples/second: 69.95
============================================================

DETAILED LINE PROFILING RESULTS
============================================================
Timer unit: 1e-09 s

Total time: 18.9137 s
File: /Users/francescocapuano/Desktop/lerobot/lerobot/common/datasets/profile_streaming_dataset.py
Function: iterate_dataset at line 171

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   171                                               def iterate_dataset(ds, n):
   172                                                   # Iterating without warmup for line profiling
   173         1       5000.0   5000.0      0.0          iterator = iter(ds)
   174         1       1000.0   1000.0      0.0          start_time = time.time()
   175      2001     447000.0    223.4      0.0          for _ in range(n):
   176      2000        2e+10    9e+06    100.0              next(iterator)
   177         1       2000.0   2000.0      0.0          end_time = time.time()
   178         1       1000.0   1000.0      0.0          return end_time - start_time

Total time: 18.908 s
File: /Users/francescocapuano/Desktop/lerobot/lerobot/common/datasets/streaming_dataset.py
Function: __iter__ at line 151

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   151                                               def __iter__(self) -> Iterator[Dict[str, torch.Tensor]]:
   152         1       1000.0   1000.0      0.0          buffer_indices_generator = self._iter_random_indices(self.rng, self.buffer_size)
   153                                           
   154                                                   # This buffer is populated while iterating on the dataset's shards
   155         1          0.0      0.0      0.0          frames_buffer = []
   156         2    1388000.0 694000.0      0.0          idx_to_iterable_dataset = {
   157                                                       idx: self._make_iterable_dataset(self.hf_dataset.shard(self.num_shards, index=idx))
   158         1       1000.0   1000.0      0.0              for idx in range(self.num_shards)
   159                                                   }
   160                                           
   161         1       1000.0   1000.0      0.0          try:
   162      3000    2542000.0    847.3      0.0              while available_shards := list(idx_to_iterable_dataset.keys()):
   163      3000   11237000.0   3745.7      0.1                  shard_key = next(self._infinite_generator_over_elements(available_shards))
   164      3000     423000.0    141.0      0.0                  dataset = idx_to_iterable_dataset[shard_key]  # selects which shard to iterate on
   165      3000        2e+10    6e+06     99.5                  for frame in self.make_frame(dataset):
   166      3000    1273000.0    424.3      0.0                      if len(frames_buffer) == self.buffer_size:
   167      2000    2520000.0   1260.0      0.0                          i = next(buffer_indices_generator)
   168      2000     711000.0    355.5      0.0                          yield frames_buffer[I]
   169      1999   67460000.0  33746.9      0.4                          frames_buffer[i] = frame
   170                                                               else:
   171      1000     311000.0    311.0      0.0                          frames_buffer.append(frame)
   172      2999    3096000.0   1032.3      0.0                      break  # random shard sampled, switch shard
   173                                           
   174         2       2000.0   1000.0      0.0          except (
   175         1          0.0      0.0      0.0              RuntimeError,
   176         1          0.0      0.0      0.0              StopIteration,
   177                                                   ):  # NOTE: StopIteration inside a generator throws a RuntimeError since 3.7
   178                                                       # Remove exhausted shard
   179                                                       del idx_to_iterable_dataset[shard_key]
   180                                           
   181                                                   # Once shards are all exhausted, shuffle the buffer and yield the remaining frames
   182                                                   self.rng.shuffle(frames_buffer)
   183                                                   yield from frames_buffer

Total time: 1e-06 s
File: /Users/francescocapuano/Desktop/lerobot/lerobot/common/datasets/streaming_dataset.py
Function: _make_iterable_dataset at line 185

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   185                                               def _make_iterable_dataset(self, dataset: datasets.Dataset) -> Iterator:
   186         1       1000.0   1000.0    100.0          return iter(dataset)

Total time: 18.8076 s
File: /Users/francescocapuano/Desktop/lerobot/lerobot/common/datasets/streaming_dataset.py
Function: make_frame at line 188

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   188                                               @profile
   189                                               def make_frame(self, dataset_iterator: datasets.IterableDataset) -> Generator:
   190                                                   """Makes a frame starting from a dataset iterator"""
   191      3000 1582055000.0 527351.7      8.4          item = next(dataset_iterator)
   192      3000   37904000.0  12634.7      0.2          item = item_to_torch(item)
   193                                           
   194                                                   # Get episode index from the item
   195      3000     532000.0    177.3      0.0          ep_idx = item["episode_index"]
   196                                           
   197                                                   # Load video frames, when needed
   198      3000   10121000.0   3373.7      0.1          if len(self.meta.video_keys) > 0:
   199      3000     457000.0    152.3      0.0              current_ts = item["timestamp"]
   200      3000    7851000.0   2617.0      0.0              query_timestamps = self._get_query_timestamps(current_ts, None)
   201      3000        2e+10    6e+06     90.6              video_frames = self._query_videos(query_timestamps, ep_idx)
   202      3000    2817000.0    939.0      0.0              item = {**video_frames, **item}
   203                                           
   204                                                   # Add task as a string
   205      3000     712000.0    237.3      0.0          task_idx = item["task_index"]
   206      3000  124168000.0  41389.3      0.7          item["task"] = self.meta.tasks.iloc[task_idx].name
   207                                           
   208      3000     469000.0    156.3      0.0          yield item


RANDOMNESS ANALYSIS
============================================================
Correlation between iteration index and frame index: 0.0772
(Correlation close to 0 indicates more random access pattern)
(Correlation close to 1 indicates sequential access pattern)
============================================================
```
- 📝 **Note:** Warmup takes time mostly due to the need to initialize the `VideoDecoder` from `torchcodec`
- 📝 **Note:** The random pattern in the index of the dataset row obtained versus the iteration index graphically illustrates the de-correlation between the two objects guaranteed by the approach presented in `streaming_dataset.py` 👇 
Screenshot 2025-05-26 at 19 33 18

**Comment by lhoestq** [comment]
Btw you can already check if it works well within a torch DataLoader with multiple workers. 

In the current state I believe some workers would shut down because you use `datasets.IterableDataset` objects with a single shard (the single shard id given to worker with rank=0, and the other ranks shut down)

File: lerobot/common/datasets/streaming_dataset.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+    def load_hf_dataset(self) -> datasets.IterableDataset:
+        dataset = load_dataset(self.repo_id, split="train", streaming=self.streaming)
+        self.streaming_from_local = False
+
+        # TODO(fracapuano): Add support for streaming from a local folder and not only from HF Hub
+        return dataset
+
+    def __iter__(self) -> Iterator[Dict[str, torch.Tensor]]:
+        buffer_indices_generator = self._iter_random_indices(self.rng, self.buffer_size)
```
[the remark this answers is not in the document]
Comment (fracapuano) [reply lerobot/common/datasets/streaming_dataset.py]: nice catch, thank you so much. I have now changed this to make it fully reproducible, so that we have a new parameter given at init which allows user to control whether they want to shuffle items between "epochs" (similarily to what is done with in-memory datasets iiuc)

File: lerobot/common/datasets/streaming_dataset.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        frames_buffer = []
+        idx_to_iterable_dataset = {
+            idx: self._make_iterable_dataset(self.hf_dataset.shard(self.num_shards, index=idx))
+            for idx in range(self.num_shards)
+        }
+
+        try:
+            while available_shards := list(idx_to_iterable_dataset.keys()):
+                shard_key = next(self._infinite_generator_over_elements(available_shards))
+                dataset = idx_to_iterable_dataset[shard_key]  # selects which shard to iterate on
```
[the remark this answers is not in the document]
Comment (fracapuano) [reply lerobot/common/datasets/streaming_dataset.py]: Love to hear this 🥹 Will open a PR with it soon if you believe it's valid

File: lerobot/common/datasets/lerobot_dataset.py [hunk 3]
```diff
@@ -151,6 +151,10 @@ def get_video_file_path(self, ep_index: int, vid_key: str) -> Path:
         fpath = self.video_path.format(video_key=vid_key, chunk_index=chunk_idx, file_index=file_idx)
         return Path(fpath)
 
+    @property
+    def url_root(self) -> str:
+        return f"https://huggingface.co/datasets/{self.repo_id}/resolve/main"
```
[the remark this answers is not in the document]
Comment (fracapuano) [reply lerobot/common/datasets/lerobot_dataset.py]: amazing, thank you. now changed 👍
